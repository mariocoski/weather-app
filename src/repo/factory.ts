
import IFactoryConfig from './FactoryConfig';
import localFactory from './local/factory';
import remoteFactory from './remote/factory';

export default ({ local, remote }: IFactoryConfig = {}) => {

  const localFacade = localFactory(local);
  const remoteFacade = remoteFactory(remote);

  return {
    local: localFacade,
    remote: remoteFacade,
  };
};