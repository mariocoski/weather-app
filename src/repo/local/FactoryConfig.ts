export interface IReading {
  readonly timestamp: number;
  readonly temp: number;
}

export interface IWeather {
  readonly data: IReading[];
  readonly loaded: boolean;
  readonly loading: boolean;
}

export enum IDataSourceEnum {
  api = 'api',
  csv = 'csv'
}

export interface IState {
  readonly countryId: string;
  readonly dataSource: IDataSourceEnum;
  readonly files: any[];
  readonly weather: IWeather;
}

export default interface IFactoryConfig {
  readonly emitChange?: () => void;
}