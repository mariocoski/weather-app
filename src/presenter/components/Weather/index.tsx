import * as React from 'react';
import { Container, Table } from 'semantic-ui-react';
import { IReading, IWeather} from '../../../repo/local/FactoryConfig';
import getDate from '../../utils/getDate';
import getTime from '../../utils/getTime';

interface IOptions {
  readonly weather: IWeather;
}

export default ({
  weather,
}: IOptions) => (<Container>
  <Table celled={true} inverted={true}>
    {weather.loaded && weather.data.length > 0 && <Table.Header>
      <Table.Row>
        <Table.HeaderCell>Date</Table.HeaderCell>
        <Table.HeaderCell>Time</Table.HeaderCell>
        <Table.HeaderCell>Temp [K]</Table.HeaderCell>
      </Table.Row>
    </Table.Header>}
    <Table.Body>
      { weather.loading && !weather.loaded && <Table.Row>
        <Table.Cell positive={true} colSpan={3}>
          Loading...
        </Table.Cell>
      </Table.Row> }
      { !weather.loading && weather.loaded && weather.data.length === 0 && <Table.Row>
        <Table.Cell warning={true}  colSpan={3}>
         No weather forecase found... :(
        </Table.Cell>
      </Table.Row> }
      { 
        !weather.loading && weather.loaded && weather.data.length > 0  &&
        weather.data.map((reading: IReading, key: number) => (<Table.Row key={key}>
          <Table.Cell>{getDate(reading.timestamp)}</Table.Cell>
          <Table.Cell>{getTime(reading.timestamp)}</Table.Cell>
          <Table.Cell >{reading.temp}</Table.Cell>
        </Table.Row>)) 
      }
    </Table.Body>
  </Table>
</Container>);
