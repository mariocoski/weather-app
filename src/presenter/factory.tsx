import * as React from 'react';
import Layout from './components/Layout';
import Provider from './components/utils/Provider';
import FactoryConfig from './FactoryConfig';

export default ({ service, observer }: FactoryConfig) => {
  return <Provider 
           service={service} 
           observer={observer} 
           render={Layout} 
         />;
};