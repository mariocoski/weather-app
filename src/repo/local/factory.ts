import FactoryConfig, { IDataSourceEnum, IState } from './FactoryConfig';

export const initialState: IState = {
  countryId: '2643743', // London
  dataSource: IDataSourceEnum.api,
  files: [],
  weather: {
    data: [],
    loaded: false,
    loading: false,
  },
};

export default ({ emitChange = () => null }: FactoryConfig = {}) => {

  let state = initialState;

  const getState = (): IState => {
    return state;
  };

  const patchState = (patch: Partial<IState>) => {
  
    state = { ...state, ...patch };
    // tslint:disable-next-line:no-console
    console.log("PATCH STATE", patch);
    emitChange();
  };
  
  const clearState = () => {
    state = {...initialState};
  };

  return { getState, patchState, clearState };
};