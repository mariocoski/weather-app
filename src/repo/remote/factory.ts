
import axiosInstance from 'axios';
import { AxiosInstance }from 'axios';
import {defaultTo} from 'lodash';
import { IDataSourceEnum } from '../local/FactoryConfig';
import IFactoryConfig from './FactoryConfig';
import getWeatherDataFromApi from './functions/getWeatherData/axios';
import getWeatherDataFromCsv from './functions/getWeatherData/csv';

const baseURL = process.env.REACT_APP_WEATHER_API_URL || 'http://api.openweathermap.org/data/2.5/forecast'; 
const defaultAxios =  async() => axiosInstance.create({ baseURL });

interface IOptions {
  readonly axios?: () => Promise<AxiosInstance>;
  readonly type: IDataSourceEnum;
  readonly files: any[];
}

const getWeatherDataFactory = (config: IOptions = {
  axios: defaultAxios,
  files: [],
  type: IDataSourceEnum.api,
}) => {
  const axios = defaultTo(config.axios, defaultAxios);
  switch(config.type){
    case IDataSourceEnum.csv: 
      return getWeatherDataFromCsv({ files: config.files });
    case IDataSourceEnum.api:
      return getWeatherDataFromApi({ axios });
  }
}

export default (config: IFactoryConfig = {
  axios: defaultAxios,
  type: IDataSourceEnum.api,
}) => {
  // tslint:disable-next-line:no-console
  return {
    getWeatherData: getWeatherDataFactory,
  }
}