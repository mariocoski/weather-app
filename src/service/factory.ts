import { IDataSourceEnum, IWeather } from '../repo/local/FactoryConfig';
import FactoryConfig from './FactoryConfig';
import getWeatherData from './functions/getWeatherData';

export default ({ repo }: FactoryConfig) => {
  return {
    clearService: () => {
      repo.local.clearState();
    },
    getDataSource: () => {
      return repo.local.getState().dataSource;
    },
    getWeather: () => {
     return repo.local.getState().weather;
    },
    getWeatherData: getWeatherData({ repo }),
    setDataSource: (dataSource: IDataSourceEnum) => {
      repo.local.patchState({
        dataSource,
      });
    },
    setFiles: (files: any) => {
      repo.local.patchState({
        files,
      });
    },
    setWeather: (weather: IWeather) => {
      return repo.local.patchState({
        weather
      });
    },
  }
}