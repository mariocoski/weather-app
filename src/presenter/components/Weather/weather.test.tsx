
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { initialState } from '../../../repo/local/factory';
import Weather from './index';

it('can match weather snapshot', () => {
  const { weather } = initialState;
 
  const component = renderer.create(
   <Weather 
     weather={weather}
   />,
  );

  const tree = component.toJSON();
 
  expect(tree).toMatchSnapshot();
});

it('should show message that no foreast is available', () => {
 
  const component = renderer.create(
   <Weather 
     weather={{
       data: [],
       loaded: true,
       loading: false,
     }}
   />,
  );

  const tree = component.toJSON();
 
  expect(tree).toMatchSnapshot();
});


it('should show loading message when loading weather forecast', () => {
  
   const component = renderer.create(
    <Weather 
      weather={{
        data: [],
        loaded: false,
        loading: true,
      }}
    />,
   );
 
   const tree = component.toJSON();
  
   expect(tree).toMatchSnapshot();
 });

