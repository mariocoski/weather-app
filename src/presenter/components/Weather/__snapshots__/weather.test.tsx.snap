// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can match weather snapshot 1`] = `
<div
  className="ui container"
>
  <table
    className="ui celled inverted table"
  >
    <tbody
      className=""
    />
  </table>
</div>
`;

exports[`should show loading message when loading weather forecast 1`] = `
<div
  className="ui container"
>
  <table
    className="ui celled inverted table"
  >
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className="positive"
          colSpan={3}
        >
          Loading...
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`should show message that no foreast is available 1`] = `
<div
  className="ui container"
>
  <table
    className="ui celled inverted table"
  >
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className="warning"
          colSpan={3}
        >
          No weather forecase found... :(
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
